ARMmbed/mbed-os
# Nordic BLE: Allow configuration of softdevice parameters

## Description

Allow configuration (via defines) of some of the key settings for the nordic softdevice.
* CENTRAL_LINK_COUNT
* PERIPHERAL_LINK_COUNT
* gatts_enable_params.attr_tab_size
* common_enable_params.vs_uuid_count
* gatts_enable_params.service_changed

These settings control the range of functionality enabled in the softdevice as well as ram consumption.
In particular, reducing these values is critical to enable usage of 16K nrf51 devices.

The defaults have not been changed so this should not effect existing users.

The values can be adjusted on a per-project basis by target_overrides in `mbed_app.json`, eg:
``` json
{
    "target_overrides": {
        "*": {
            "nordic-ble.central_link_count": 0,
            "nordic-ble.peripheral_link_count": 1,
            "nordic-ble.gatt_attr_tab_size": "0x4A0",
            "nordic-ble.uuid_table_max_entries": 2,
            "nordic-ble.is_srvc_changed_charact_present": 1
        },
}
```

Note: There are a couple of minor edits to the logging lines in `mbed-os/targets/TARGET_NORDIC/TARGET_NRF5/TARGET_SDK11/softdevice/common/softdevice_handler/softdevice_handler.c`

These edits were not necessary in any later versions of the SDK.

## Status

**READY**

## Migrations

There should be no required changes to end user code.

## Related PRs

This is a cleaned up re-submission of #5980

## Steps to test or reproduce

Changing the target config values above in defines/config should result in a change of ram requirements.

When adjusting settings it is useful to enable `target.nrf_enable_logging` as shown above.  You will then get a printout on the mbed stdout serial port during softdevice configuration as such: 
```
sd_ble_enable: RAM START at 0x20002ef8
sd_ble_enable: RAM start should be adjusted to 0x20001f18
RAM size should be adjusted to 0x60e8
```

If the START value is lower than the RAM ORIGIN in the current linker file, the linker file can be copied into your project and edited to set the RAM ORIGIN and LENGTH to the printed values.
The mbed compile command can then have the linker arg set to your local linker file.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [x] Feature
    [ ] Breaking change
